core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -231,6 +232,25 @@ abstract class AbstractFetcherThread(name: String,
     handlePartitionsWithErrors(partitionsWithError)
   }
 
+  def markPartitionsForTruncation(partitionAndOffsets: Map[TopicPartition, Long]) {
+    partitionMapLock.lockInterruptibly()
+    try {
+      val newStates = partitionStates.partitionStates.asScala.map { state =>
+        val maybeNeedTruncation = partitionAndOffsets.get(state.topicPartition()) match {
```

Comment:
Yeah I can actually change this method to iterate `partitionAndOffsets`, do lookups in `partitionAndOffsets`, and use `partitionStates.update(Map<TopicPartition, S>)` to update partitionStates.

The advantage of the suggested approach is that it will increase the performance of this particular method. But overall this is not going to change the performance of ReplicaAlterLogDirsThread because ReplicaAlterLogDirsThread.buildFetchRequest needs to iterate the entire partitionMap in O(n) time anyway.

The advantage of the current approach is that, the implementation of this method is very similar to `AbstractFetcherThread.markTruncationComplete()`, thus making the code easier to maintain in the long term.

Thus there are pros and cons of each approach. I can make the suggested change if you prefer. What do you think?

Reply:
Thanks. I realized that the signature of this method could be simplified to `markPartitionsForTruncation(topicPartition: TopicPartition)`. I have updated the patch to do the following to optimize the performance of this method.

```
Option(partitionStates.stateValue(topicPartition)).foreach { state =>
  val newState = PartitionFetchState(state.fetchOffset, state.delay, truncatingLog = true)
  partitionStates.updateAndMoveToEnd(topicPartition, newState)
}
```